pkg/user/generator.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2016 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// For how the uidshift and uidcount are generated please check:
+// http://cgit.freedesktop.org/systemd/systemd/commit/?id=03cfe0d51499e86b1573d1
+
+package user
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/coreos/rkt/pkg/group"
+	"github.com/coreos/rkt/pkg/passwd"
+	"github.com/hashicorp/errwrap"
+)
+
+// Generator defines the interface for generating a uid/gid.
+type Generator interface {
+	IDs() (int, int, error)
+}
+
+// PickGenerator picks the first non-nil generator and returns it.
+// If no non-nil generator was found, nil is returned.
+func PickGenerator(gs ...Generator) Generator {
+	for _, g := range gs {
+		if g != nil {
+			return g
+		}
+	}
+
+	return nil
+}
+
+type fromEtc struct {
+	rootPath string
+	username string
+	group    string
+}
+
+// FromEtc returns a new uid/gid generator by parsing etc/passwd, and etc/group
+// relative from the given rootPath looking for the given username, or group.
+// If username is empty string the etc/passwd lookup will be ommitted.
+// If group is empty string the etc/group lookup will be ommitted.
+func FromEtc(rootPath, username, group string) Generator {
+	return fromEtc{
+		rootPath: rootPath,
+		username: username,
+		group:    group,
+	}
+}
+
+func (e fromEtc) IDs() (uid int, gid int, err error) {
+	uid, gid = -1, -1
+
+	if uid, err = passwd.LookupUidFromFile(
+		e.username,
+		filepath.Join(e.rootPath, "etc/passwd"),
+	); e.username != "" && err != nil {
+		return
+	}
+
+	if gid, err = group.LookupGidFromFile(
+		e.group,
+		filepath.Join(e.rootPath, "etc/group"),
+	); e.group != "" && err != nil {
```

Comment:
Please don't use multiple line statements in `if` initializations.
```go
gid, err := .....
if e.group != "" && err != nil {
  ...
}

return
```
would be much better since you are using named return values.